[LLDB] Add type casting to DIL.

cmtice opened the pull request:
This adds basic c-style type casting to DIL. It handles basic built-in types: bool, char, short, int, long, double, float, and void. It also handles user-defined types (e.g. class names, typedefs, etc.). It does NOT handle C++ templates at the moment. Not sure if this is something we would want to add to DIL later or not. DIL will not be supporting C++-specific-style casts (static_cast<>, reinterpret_cast<>, or dynamic_cast<>).

This PR adds quite a bit of type-parsing, which must happen in the parser, to resolve ambiguity ("is this thing I'm looking at a type cast or just an expression in parentheses?").

I'd be happy to break this up into smaller PRs if someone could suggest a reasonable way to break it up.

cmtice committed:
[LLDB] Add type casting to DIL.

This adds basic c-style type casting to DIL. It handles basic
built-in types: bool, char, short, int, long, double, float, and void.
It also handles user-defined types (e.g. class names, typedefs, etc.).
It does NOT handle C++ templates at the moment. Not sure if this is
something we would want to add to DIL later or not. DIL will not
be supporting C++-specific-style casts (static_cast<>, reinterpret_cast<>,
or dynamic_cast<>).

This PR adds quite a bit of type-parsing, which must happen in the
parser, to resolve ambiguity ("is this thing I'm looking at a type cast
or just an expression in parentheses?").

I'd be happy to break this up into smaller PRs if someone could suggest
a reasonable way to break it up.

cmtice committed:
Merge remote-tracking branch 'origin/main' into dil-type-case

kuilpd commented on lldb/test/API/commands/frame/var-dil/expr/CStyleCast/TestFrameVarDILCStyleCast.py:
```diff
@@ -0,0 +1,233 @@
+          Is32Bit = True;
+
+        if Is32Bit:
+          self.expect("frame variable '(int)arr'", error=True,
+                      substrs=["cast from pointer to smaller type 'int' loses"
+                               " information"])
```
Shouldn't this success on a 32-bit system?

kuilpd commented on lldb/test/API/commands/frame/var-dil/expr/CStyleCast/main.cpp:
```diff
@@ -0,0 +1,81 @@
+  float fmax = std::numeric_limits<float>::max();
+  float fdenorm = std::numeric_limits<float>::denorm_min();
+
+  // TestCStyleCastBuiltins
+  // TestCStyleCastBasicType
+  // TestCStyleCastPointer
+  // TestCStyleCastNullptrType
+  if (false) { // Set a breakpoint here
+  }
```
```suggestion
```

kuilpd commented on lldb/test/API/commands/frame/var-dil/expr/CStyleCast/main.cpp:
```diff
@@ -0,0 +1,81 @@
+  int arr_1d[] = {1, 2, 3, 4};
+  int arr_2d[2][3] = {{1, 2, 3}, {4, 5, 6}};
+
+  // TestCStyleCastArray
+  // TestCStyleCastReference
```
```suggestion
```

kuilpd commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -608,4 +635,238 @@ Interpreter::Visit(const BooleanLiteralNode *node) {
+  }
+
+  switch (cast_kind) {
+  case CStyleCastKind::eEnumeration: {
```
There are so many type checks between this, `VerifyCStyleCastType` and `CastToBasicType`, which all seem to check similar things. I haven't looked too much into these yet, but is there any way we can simplify all this? My only idea would be to unify `VerifyCStyleCastType` and this switch here, since we don't do checks separately like in `lldb-eval` anymore, while also removing the repeated checks that are done in `CastToBasicType` (if any). But I'm not sure if this would look any better or not.

cmtice replied:
It looks like there were some redundant type checks; I've eliminated them and consolidated the code as much as I can.

kuilpd commented on lldb/test/API/commands/frame/var-dil/expr/CStyleCast/TestFrameVarDILCStyleCast.py:
```diff
@@ -0,0 +1,233 @@
+class TestFrameVarDILCStyleCast(TestBase):
+    NO_DEBUG_INFO_TESTCASE = True
+
+    def expect_var_path(self, expr, compare_to_framevar=False, value=None, type=None):
+        value_dil = super().expect_var_path(expr, value=value, type=type)
+        if compare_to_framevar:
+            self.runCmd("settings set target.experimental.use-DIL false")
+            value_frv = super().expect_var_path(expr, value=value, type=type)
+            self.runCmd("settings set target.experimental.use-DIL true")
+            self.assertEqual(value_dil.GetValue(), value_frv.GetValue())
```
We don't need this anymore, since these new expressions are not handled by the legacy frame var.

kuilpd reviewed (commented):
I think we need to get rid of everything that references C++ keywords directly. I don't think we need to parse the builtin type manually, so we can just remove `IsSimpleTypeSpecifierKeyword`, `HandleSimpleTypeSpecifier`, and `TypeDeclaration` class altogether. Instead, parse a sequence of identifiers as a single string and call `TypeSystem::GetBuiltinTypeByName` to check if such a builtin type exists, it already has pretty much all variations of builtin types in `TypeSystemClang`. This way other languages can specify their own builtin types as well.

The algorithm of `ParseTypeId` would be something like this:
1. `ParseTypeSpecifierSeq`, that returns a string that is either a builtin type (`"unsigned long long"`) or a user defined type (`"ns::myint"`)
2. Call `ResolveTypeByName` on the string, which first attempts a `GetBuiltinTypeByName` and immediately returns a `CompilerType` if found, then searches in modules for user types
3. Handle * and & like before

I did a quick test with this logic, seems to be working.

cmtice committed:
Address review comments in cast test & fix clang-format issues.

cmtice committed:
Remove some redundant type checks. Clean up code slightly.

cmtice committed:
Fix more clang-format issues.

cmtice committed:
Fix formatting.

kuilpd reviewed (commented):
Looks much better, thank you!

What about my suggestion to use `TypeSystem::GetBuiltinTypeByName` instead of manual parsing into `TypeDeclaration` class? Do you think it's worth trying or is it up for a debate?

cmtice committed:
Remove commented out code & extra ';'
Update ResolveTypeByName to return immediately when full match is found.
Remove 'DIL' prefix from GetTypeSystemFromCU.

cmtice commented:
> I think we need to get rid of everything that references C++ keywords directly. I don't think we need to parse the builtin type manually, so we can just remove `IsSimpleTypeSpecifierKeyword`, `HandleSimpleTypeSpecifier`, and `TypeDeclaration` class altogether. Instead, parse a sequence of identifiers as a single string and call `TypeSystem::GetBuiltinTypeByName` to check if such a builtin type exists, it already has pretty much all variations of builtin types in `TypeSystemClang`. This way other languages can specify their own builtin types as well.
> 
> The algorithm of `ParseTypeId` would be something like this:
> 
> 1. `ParseTypeSpecifierSeq`, that returns a string that is either a builtin type (`"unsigned long long"`) or a user defined type (`"ns::myint"`)
> 2. Call `ResolveTypeByName` on the string, which first attempts a `GetBuiltinTypeByName` and immediately returns a `CompilerType` if found, then searches in modules for user types
> 3. Handle * and & like before
> 
> I did a quick test with this logic, seems to be working.

The biggest problem, is in your step 2. We can't make GetBuiltinTypeByName the first thing in ResolveTypeByName. Doing that would prevent users from every being able to redefine types with the same names as the builtins, which is something I've been told is allowed in Swift, e.g.

cmtice commented:
I have been wondering: If people think this PR is too big to review, I *could* pull most of the type parsing out into a separate PR and have that reviewed by itself...

What do folks think?

kuilpd commented:
> The biggest problem is in your step 2. We can't make GetBuiltinTypeByName the first thing in ResolveTypeByName. Doing that would prevent users from every being able to redefine types with the same names as the builtins, which is something I've been told is allowed in Swift, e.g.

Ah, I see, I didn't realize that this logic might not work for other languages. Mainly I suggested to change the way we parse the basic types, using `GetBuiltinTypeByName` instead of doing it manually. But this manual implementation comes from lldb-eval, which handled only C++, so it prioritizes the basic types the same way: if it finds something that looks like a basic type, it just returns the according CompilerType without looking for user types any further. 

In any case, Swift, if I understand correctly, doesn't really have builtin types, all the types like `Int` or `Float` are effectively user types defined in the standard library. The important distinction is that C++ builtin types may not be found by `ResolveTypeByName` that searches types in modules, so we need explicitly check if something is a builtin type, while for Swift we don't need to. So from my understanding `GetBuiltinTypeByName` in `TypeSystemSwift` can just always return nothing, and the code can move on to searching user types.

Need some Swift insight here :) @jimingham @adrian-prantl

kuilpd commented:
> Swift does have builtin types, they just aren't exposed to the user directly. 

@jimingham 
Ah, I see. Well, then we can still check for builtin types first, and when it comes to implementing `GetBuiltinTypeByName` in `TypeSystemSwift` for DIL, you can decide whether you want to allow users to lookup builtin types or leave that function empty.

cmtice committed:
Simplify parsing of builtin type names.

Following a  reviewer suggestion. This eliminates the need for the
TypeSpecifier & TypeDeclaration classes which simplifies things
a bit.

cmtice commented:
I've simplified the type parsing a bit, as suggested above (for the builtin types), which eliminated ~400 lines of code (woo hoo!).  Verified that this still passes all the type-casting tests.

cmtice committed:
Merge remote-tracking branch 'origin/main' into dil-type-case

cmtice commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -44,6 +47,31 @@ nested_name_specifier = type_name "::"
+               | "unsigned"
+               | "float"
+               | "double"
+               | "void" ;
```
That is in fact the way the (updated) implementation works: Just expecting identifiers and using calls to LLDB's type system to determine if the name is for a builtin type or not. I just forgot to update the grammar here when I changed the implementation to work that way. I will update the grammar (thanks for catching this).

cmtice committed:
Update grammar to not explicitly name builtin types.
